testonly/integration/etcd/etcd.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/embed"
+)
+
+// MaxEtcdStartAttempts is the max number of start attempts made before it fails.
+const MaxEtcdStartAttempts = 3
+
+// StartEtcd returns a started, ready to use embedded etcd, along with a client and a cleanup
+// function (that must be defer-called). There's no need to defer-close etcd of client, cleanup
+// closes both.
+//
+// A temp directory and random ports are used to setup etcd.
+func StartEtcd() (e *embed.Etcd, c *clientv3.Client, cleanup func(), err error) {
+	var dir string
+	dir, err = ioutil.TempDir("", "etcdquota-test-")
+	if err != nil {
+		return
+	}
+
+	cleanup = func() {
+		if c != nil {
+			c.Close()
+		}
+		if e != nil {
+			e.Close()
+		}
+		os.RemoveAll(dir)
+	}
+
+	for i := 0; i < MaxEtcdStartAttempts; i++ {
+		e, err = tryStartEtcd(dir)
+		if err == nil {
+			break
+		}
+		if strings.Contains(err.Error(), "address already in use") {
+			continue
+		}
+		cleanup()
+		return
+	}
+	if e == nil {
+		cleanup()
+		err = fmt.Errorf("failed to start etcd: too many attempts")
+		return
+	}
+
+	select {
+	case <-e.Server.ReadyNotify():
+		// OK
+	case <-time.After(5 * time.Second):
+		cleanup()
+		err = errors.New("timed out waiting for etcd")
+		return
+	}
+
+	c, err = clientv3.New(clientv3.Config{
+		Endpoints:   []string{e.Config().LCUrls[0].String()},
+		DialTimeout: 5 * time.Second,
+	})
+	if err != nil {
+		cleanup()
+	}
+	return
+}
+
+func tryStartEtcd(dir string) (*embed.Etcd, error) {
+	p1 := 1000 + rand.Intn(4000) // Pick a port in (1000,5000]. The interval is arbitrary.
```

Comment:
If you mean `integration.listen()`, the part that we could reuse is basically `net.Listen("tcp", ":0")`. I've reworked this to use `net.Listen()`.